### PR TITLE
upgrade Kamon to 2.1.12

### PR DESCRIPTION
**Pull request by bcarter97**
## Description

- Upgrade Kamon to 2.1.12 to remove the `nanohttpd` peer dependency

### Migration notes

- [Implicit Config import is no longer needed.](https://kamon.io/docs/latest/guides/migration/from-1.x-to-2.0/#there-is-a-new-kamoninit-method) Instead, `Kamon.init()` is called early in the application which bootstraps the metric reporting
- [`stopAllReports` has been renamed -> `stopAllModules`](https://github.com/kamon-io/kamon-prometheus/issues/30#issuecomment-510873694)
- [`refine` has been renamed -> `withTags`](https://kamon.io/docs/latest/guides/migration/from-1.x-to-2.0/#metrics-are-now-tagged-instead-of-refined)
- [Prometheus](https://prometheus.io/) has been added to the docker-compose

### CODE REVIEW

**Comment by btanyab** [comment]
In the [application.conf](https://github.com/sky-uk/kafka-message-scheduler/blob/eb7d2b04266913552b1aa20cc402cb60161455cb/scheduler/src/main/resources/application.conf#L53) configuration path need to be changed to : 

```scala 
kamon {
  environment.service = "kafka-message-scheduler"
  instrumentation.akka.filters {
    "actors.track" {
      includes = ["kafka-message-scheduler/user/scheduling-actor", "kafka-message-scheduler/user/publisher-actor"]
      excludes = ["kafka-message-scheduler/system/**"]
    }
  }
```
as per [Kamon docs](https://kamon.io/docs/latest/instrumentation/akka/metrics/#filtered-metrics)

**Comment by peter-hazell** [comment]
> In the [application.conf](https://github.com/sky-uk/kafka-message-scheduler/blob/eb7d2b04266913552b1aa20cc402cb60161455cb/scheduler/src/main/resources/application.conf#L53) configuration path need to be changed to :
> 
> ```scala
> kamon {
>   environment.service = "kafka-message-scheduler"
>   instrumentation.akka.filters {
>     "actors.track" {
>       includes = ["kafka-message-scheduler/user/scheduling-actor", "kafka-message-scheduler/user/publisher-actor"]
>       excludes = ["kafka-message-scheduler/system/**"]
>     }
>   }
> ```
> 
> as per [Kamon docs](https://kamon.io/docs/latest/instrumentation/akka/metrics/#filtered-metrics)

Good spot Tanya - final paragraph of the migration guide explains as well: https://kamon.io/docs/latest/guides/migration/from-1.x-to-2.0/

**Comment by peter-hazell** [comment]
@bcarter97 I think there are couple of parts of the migration guide we should revisit, even though we did see metrics being produced when we tested this

https://kamon.io/docs/latest/guides/migration/from-1.x-to-2.0/#there-is-a-new-kamoninit-method - `Kamon.init()` needs adding in our main method

https://kamon.io/docs/latest/guides/migration/from-1.x-to-2.0/#there-is-a-new-kamoninit-method - I think there is an aspectj dependency in here at the mo, that can be changed to kanela.

**Comment by btanyab** [comment]
Would it be easier to import  just `kamon-bundle` (instead of each module separately) which bundles all the instrumentation modules available in Kamon and the Kanela agent

File: scheduler/src/main/resources/application.conf [hunk 1]
```diff
@@ -52,11 +52,9 @@ akka {
...
 kamon {
   environment.service = "kafka-message-scheduler"
-  util.filters {
-    "akka.tracked-actor" {
-      includes = ["kafka-message-scheduler/user/scheduling-actor", "kafka-message-scheduler/user/publisher-actor"]
-      excludes = ["kafka-message-scheduler/system/**"]
-    }
+  instrumentation.akka.filters.actors.track {
+    includes = ["kafka-message-scheduler/user/scheduling-actor", "kafka-message-scheduler/user/publisher-actor"]
+    excludes = ["kafka-message-scheduler/system/**"]
```
Comment (bcarter97) [inline scheduler/src/main/resources/application.conf:57]: fixes https://github.com/sky-uk/kafka-message-scheduler/pull/101#issuecomment-800134794

File: build.sbt [hunk 2]
```diff
@@ -27,10 +27,9 @@ val dependencies = Seq(
   "org.codehaus.janino"        % "janino"                       % "3.0.13" % Runtime,
   "com.github.pureconfig"      %% "pureconfig"                  % pureConfigVersion,
   "com.github.pureconfig"      %% "pureconfig-cats"             % pureConfigVersion,
-  "io.kamon"                   %% "kamon-prometheus"            % "1.1.2",
-  "io.kamon"                   %% "kamon-akka-2.5"              % "1.1.4",
-  "io.kamon"                   %% "kamon-core"                  % "1.1.6",
-  "io.kamon"                   %% "kamon-jmx-collector"         % "0.1.8",
+  "io.kamon"                   %% "kamon-bundle"                % "2.1.12",
+  "io.kamon"                   %% "kamon-prometheus"            % "2.1.12",
+  "io.kamon"                   %% "kamon-jmx-collector"         % "1.0.0-RC1",
```
Comment (bcarter97) [inline build.sbt:32]: kamon bundle doesn't seem to contain prometheus, the build will fail without it.

File: build.sbt [hunk 3]
```diff
@@ -60,7 +59,7 @@ lazy val dockerSettings = Seq(
   dockerUpdateLatest := updateLatest.value,
   dockerCommands ++= Seq(
     Cmd("USER", "root"),
-    Cmd("RUN", "apk update && apk add bash")
+    Cmd("RUN", "apk update && apk add bash && apk add eudev")
```
Comment (bcarter97) [inline build.sbt:62]: eudev fixes a large console warn in the build step.

File: docker/docker-compose.yml [hunk 4]
```diff
@@ -27,3 +27,12 @@ services:
...
     - "9095:9095"
+
+  prometheus:
+    image: prom/prometheus:v1.7.2
+    ports:
+      - "9090:9090"
+    volumes:
+      - ./prometheus.yml:/etc/prometheus/prometheus.yml
+    command:
+      - '-config.file=/etc/prometheus/prometheus.yml'
```
Comment (bcarter97) [inline docker/docker-compose.yml:38]: prometheus is mentioned in the README - this provides the basic implementation.

File: build.sbt [hunk 5]
```diff
@@ -27,10 +27,9 @@ val dependencies = Seq(
   "org.codehaus.janino"        % "janino"                       % "3.0.13" % Runtime,
   "com.github.pureconfig"      %% "pureconfig"                  % pureConfigVersion,
   "com.github.pureconfig"      %% "pureconfig-cats"             % pureConfigVersion,
-  "io.kamon"                   %% "kamon-prometheus"            % "1.1.2",
-  "io.kamon"                   %% "kamon-akka-2.5"              % "1.1.4",
-  "io.kamon"                   %% "kamon-core"                  % "1.1.6",
-  "io.kamon"                   %% "kamon-jmx-collector"         % "0.1.8",
+  "io.kamon"                   %% "kamon-bundle"                % "2.1.12",
+  "io.kamon"                   %% "kamon-prometheus"            % "2.1.12",
```
Comment (paulaylingdev) [inline build.sbt:31]:
```suggestion
  "io.kamon"                   %% "kamon-bundle"                % kamonVersion,
  "io.kamon"                   %% "kamon-prometheus"            % kamonVersion,
```
Lets create a val for the version as I believe both modules are created from the same group and use the same versioning.

File: build.sbt (hunk 3 above)
Comment (paulaylingdev) [inline build.sbt:62]:
```suggestion
    Cmd("RUN", "apk update && apk add bash eudev")
```

Can just list the packages after add.

Reference: https://wiki.alpinelinux.org/wiki/Alpine_Linux_package_management#Add_a_Package

File: docker/docker-compose.yml [hunk 6]
```diff
@@ -27,3 +27,12 @@ services:
       SCHEDULE_TOPICS: schedules
     ports:
     - "9095:9095"
+
+  prometheus:
+    image: prom/prometheus:v1.7.2
```
Comment (paulaylingdev) [inline docker/docker-compose.yml:32]: Quite an old verison now (Sept 2017), maybe worth using a newer one?

File: scheduler/src/main/scala/com/sky/kms/Main.scala [hunk 7]
```diff
@@ -9,11 +9,12 @@ import pureconfig.loadConfigOrThrow
 import pureconfig.module.cats._
 
 object Main extends App with LazyLogging with AkkaComponents {
-
-  val conf = loadConfigOrThrow[AppConfig]
+  SchedulerApp.metricsInit
```
Comment (paulaylingdev) [inline scheduler/src/main/scala/com/sky/kms/Main.scala:12]: Lets put this after the below log line incase it throws an exception?

**Review by paulaylingdev — COMMENTED** [review]
Minor comments but looks good!

File: docker/docker-compose.yml (hunk 6 above)
Comment (bcarter97) [reply docker/docker-compose.yml]: upgraded to 2.25.0 <- latest.

**Review by btanyab — APPROVED** [review]

**Comment by paulaylingdev** [comment]
Can we update the monitoring paragraph a little in the README.md. I think the link to 'kamon prometheus reporter' is old.

Might be worth adding a sentence that prometheus is included with the docker compose as an example.

**Review by paulaylingdev — APPROVED** [review]

**Pull request merged**